One-click Redis benchmarking and optimization (#455)
microsoft/MLOS: mlos_bench/README.md

To get started, we can adapt an example configuration to test out running `mlos-bench`.
For these instructions, we will be using Azure for our resources.

### 1. Install and activate the conda environment

From here onwards we assume we are in the project root directory.
Ensure you have a conda environment (`mlos`) set up for executing `mlos_bench`.
Create and activate the environment with:

```sh
conda env create -f conda-envs/mlos.yml
conda activate mlos
```

> 

Reviewer: ```suggestion

> Note: if you are running inside the devcontainer, this should be done automatically.

```

### 2. Make sure that you have Azure CLI tool installed and working

> > Installation instructions for `az` (Azure CLI) [can be found here](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli).

Reviewer: ```suggestion
> Installation instructions for `az` (Azure CLI) [can be found here](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli).

> Note: `az` comes preinstalled inside the devcontainer.
```

If necessary, login to Azure and set your default subscription:

> ```powershell

Reviewer: ```suggestion
```shell
```

# If using az cli for the first time, a login will be required:
az login

> # Make sure that GSL subscription is your default:

Reviewer: ```suggestion
# Make sure to set your default subscription, RG, and Storage Account for these experiments.
# For instance:
az account set --subscription "My Subscription Name"
az config set defaults.group=MyRG --local
az config set storage.account=MyStorageAccount --local 
```

az account set --subscription "..."
```

### 3. Generate access tokens to interact with Azure resources

> A script at `./scripts/generate-azure-credentials-config.ps1` produces a JSON with Azure credentials.

Reviewer: Future work: I can probably consolidate those script files so a single `scripts/generate-azure-credentials-config` without an extension will work across both platforms.  We already do this for the devcontainer stuff.

Reviewer: I did that here.

This data is in the format that can be used by our framework.

> ```powershell

Reviewer: ```suggestion
If using Windows, use the following:
```powershell
```

> # If using for the az cli first time, a login will be required

Reviewer: ```suggestion
```

Already handled in the previous step.

./scripts/generate-azure-credentials-config.ps1 > ./global_config_azure.json
```

> On Linux, use `./scripts/generate-azure-credentials-config.sh` (requires `az` and `jq` to be installed).

Reviewer: ```suggestion
On Linux (or inside the devcontainer), use the following:
```sh
./scripts/generate-azure-credentials-config.sh > ./global_config_azure.json
\`\`\` # FIXME: Remove backslash escapes in actual code - can't see to do that in the web editor

> Note: Requires `jq` to also be installed (comes preinstalled in the devcontainer).

```

### 4. Create a JSON config with DB credentials (Optional)

If you plan to store the information about experiments and benchmarks in a (remote) database like PostgreSQL or MySQL, create a JSON/JSONC file with the DB hostname and password.

> See [`mysql.jsonc`](./mlos_bench/config/storage/mysql.jsonc) or [`postgresql.jsonc`](./mlos_bench/config/storage/postgresql.jsonc) configuration files for the full list of DB parameters.

Reviewer: ```suggestion
See [`mysql.jsonc`](./mlos_bench/config/storage/mysql.jsonc) or [`postgresql.jsonc`](./mlos_bench/config/storage/postgresql.jsonc) configuration files for examples with a more complete list of DB parameters supported by underlying the [SqlAlchemy](https://www.sqlalchemy.org/library.html#reference) library.
```

Save your config in `./global_config_storage.jsonc` file.
It should look like this:

```jsonc
{
    "host": "mysql-db.mysql.database.azure.com",
    "password": "database_password"
}
```

Any parameter that is not specified in `./global_config_storage.json` will be taken from the corresponding DB's config file, e.g., [`postgresql.jsonc`](./mlos_bench/config/storage/postgresql.jsonc).

For database like SQLite or DuckDB, there is no need for an additional config file.
The data will be stored locally in a file, e.g., `./mlos_bench.duckdb`.

> See [`sqlite.jsonc`](./mlos_bench/config/storage/sqlite.jsonc) or [`duckdb.jsonc`](./mlos_bench/config/storage/duckdb.jsonc) for more details.

Reviewer: ```suggestion
See [`sqlite.jsonc`](./mlos_bench/config/storage/sqlite.jsonc) or [`duckdb.jsonc`](./mlos_bench/config/storage/duckdb.jsonc) for more details.

> Note: if no storage is specified, a basic sqlite config will be used by default.
```

?

### 5. Create a top-level configuration file for your MLOS setup

We provide a few examples of such files in [`./mlos_bench/config/cli/`](./mlos_bench/config/cli).
For example, [`azure-redis-opt.jsonc`](./mlos_bench/config/cli/azure-redis-opt.jsonc) is a configuration for optimizing Redis VM on Azure and saving the results in a local SQLite database.
Likewise, [`azure-redis-bench.jsonc`](./mlos_bench/config/cli/azure-redis-bench.jsonc) is a setup to run a single Redis benchmark (and, again, save the results in SQLite).

CLI configs like those are meant to connect several MLOS components together, namely:

- Benchmarking environment (configured in [`environments/root/root-azure-redis.jsonc`](./mlos_bench/config/environments/root/root-azure-redis.jsonc));
- Optimization engine ([`optimizers/mlos_core_flaml.jsonc`](./mlos_bench/config/optimizers/mlos_core_flaml.jsonc));
- Storage for experimental data ([`storage/sqlite.jsonc`](./mlos_bench/config/storage/sqlite.jsonc));

They also refer to other configs, e.g.

- Reusable config snippets in `"config_path"` section, and
- Additional config files containing sensitive data like DB passwords and Azure credentials.

> > Make sure that the files `./global_config_azure.json` and `./global_config_storage.json` you created in steps 3 and 4 are included in the `"globals"` section of your CLI config.

Reviewer: ```suggestion
> Make sure that the files `./global_config_azure.json` and `./global_config_storage.json` you created in the previous steps are included in the `"globals"` section of your CLI config.
```

Just a maintenance tweak in case we renumber things.

For the purpose of this tutorial, we will assume that we reuse the existing [`azure-redis-bench.jsonc`](./mlos_bench/config/cli/azure-redis-bench.jsonc) and [`azure-redis-opt.jsonc`](./mlos_bench/config/cli/azure-redis-opt.jsonc) configurations without any changes.
In a more realistic scenario, however, you might need to change and/or create new config files for some parts of your benchmarking environment.
We'll give more details on that below.

### 5. Create another config file for the parameters specific to your experiment

Copy one of our examples, e.g., [`experiment_RedisBench.jsonc`](./mlos_bench/config/experiments/experiment_RedisBench.jsonc) and name it after your experiment, e.g. `experiment_MyBenchmark.jsonc`.

In that file, you can specify any parameters that occur in your other configs, namely in `"const_args"` section of the Environment configs, or in `"config"` sections of your Service, Storage, or Optimizer configurations.

> A general rule is that the parameters from the global configs like `./global_config_azure.json` or `experiment_MyAppBench.jsonc` override the corresponding parameters in other configurations.

> > That allows us to propagate the values of the parameters that are specific to the experiment into other components of the framework and keep the majority of the config files in our library immutable and reusable.

Reviewer: This is a really important detail to document, thank you!

We might also want to include this in the `configs/README.md` section: basically just some general overview of how content flows from one config to another so that people can have some principles by which to construct their own configs reasonably well without reading the code.

Reviewer: Added this.

### 6. Run the benchmark

Now we can run our configuration with `mlos_bench`:

> ```sh

Reviewer: ```suggestion
```shell
```

In general, anywhere where there's a non-platform specific syntax being used, I try to use `shell` as the code block style identifier.

Note that using the `--globals` command line option is the same as adding `experiment_MyBenchmark.jsonc` to the `"globals"` section of the CLI config.
Same applies to all other CLI parameters - e.g., you can change the log level by adding `--log_level INFO` to the command line.

Also, note that you don't have to provide full path to the `experiment_MyBenchmark.jsonc` file - the application will look for it in the paths specified in the `"config_path"` section of the CLI config.

## Optimization

Searching for an optimal set of tunable parameters is very similar to running a single benchmark.
All we have to do is specifying the [`Optimizer`](./mlos_bench/optimizers/) in the top-level configuration, like in our [`azure-redis-opt.jsonc`](./mlos_bench/config/cli/azure-redis-opt.jsonc) example.

```sh
mlos_bench --config "./mlos_bench/mlos_bench/config/cli/azure-redis-opt.jsonc" --globals "experiment_MyBenchmark.jsonc --max_iterations 10"
```

> Note that again we use the command line option `--max_iterations` to override the default value from [`mlos_core_flaml.jsonc`](./mlos_bench/config/optimizers/mlos_core_flaml.jsonc).

Reviewer: Great example and explanation!
